src/kaleidoscope/plugin/Heatmap.h

```diff
@@ -34,16 +35,40 @@ class Heatmap : public LEDMode {
   EventHandlerResult onKeyswitchEvent(Key &mapped_key, byte row, byte col, uint8_t key_state);
   EventHandlerResult beforeEachCycle();
 
- protected:
-  void update(void) final;
+  // This class' instance has dynamic lifetime
+  //
+  class DynamicLEDMode : public LEDMode {
+   public:
 
- private:
-  static uint16_t heatmap_[ROWS][COLS];
-  static uint16_t highest_;
-  static uint32_t next_heatmap_comp_time_;
+    // Please note that storing the parent ptr is only required
+    // for those LED modes that require access to
+    // members of their parent class. Most LED modes can do without.
+    //
+    DynamicLEDMode(const Heatmap *parent);
 
-  static void shiftStats(void);
-  static cRGB computeColor(float v);
+    void resetMap();
+    EventHandlerResult onKeyswitchEvent(Key &mapped_key, byte row, byte col, uint8_t key_state);
+    EventHandlerResult beforeEachCycle();
+
+   protected:
+
+    virtual void update() final;
+
+   private:
+
+    const Heatmap *parent_;
+
+    uint16_t heatmap_[ROWS][COLS];
+    uint16_t highest_;
+    uint32_t next_heatmap_comp_time_;
```

Comment:
I'm not sure that the Heatmap plugin should operate this way. It builds its heatmap data over time, as the user types, but if it only exists while it is the active LED mode, then a user needs to have the heatmap turned on all the time in order to work. It seems more useful to me if I can have the LEDs off, or switch between modes, and still have the heatmap accumulating data while I type. Then I could switch to the Heatmap mode and see what it looks like every once in a while. @algernon – what do you think?

Reply:
Right, I didn't think of it that way. We might better stick with the original (static) version of heatmap. Going  with a dynamic LED mode where the dynamic part only stores `next_heatmap_comp_time_` would save three bytes of permanent RAM (four bytes for `next_heatmap_comp_time_` saved minus one byte for `led_mode_id_` to enable the plugin to access its dynamic LED mode).

Let's see what @algernon says.

Reply:
Yeah, it sucks that it works out this way for one of the bigger LED modes, but if the heatmap loses its data when a different mode is activated, it becomes useless for someone who wants to see the heatmap, but finds it distracting to have the LEDs on while typing.

I can think of a compromise. If Heatmap adds a boolean that controls whether the LEDs are turned on or not, it would be possible to toggle them off without turning off the dynamic LED mode. It's not a great compromise, because it wouldn't be possible for a user to turn on (for example) ActiveModColor and still have Heatmap collecting data.

Reply:
Another idea is to have two different versions of Heatmap: one dynamic, for people who need the RAM, but don't mind if it loses its data when another mode is active, and one static, for people who can spare the RAM and want it collecting data all the time.

Reply:
Those are both very nice ideas! For the first one we could add some sort of signal while it is recording, lets say one of the keys blinking red like a camera would do.

The second idea is also feasible. We could implement the static version of the plugin as a simple wrapper for the `LEDMode` derived class that the dynamic version exports. Thus, we could avoid code duplication.

Reply:
FWIW, I suggested the red blinking light because I find it sometimes difficult to keep track of which of the LED modes is active. That's difficult with those that only light up LEDs as a reaction to typing.

Reply:
I don't think having two versions of Heatmap is worth the effort, even if that effort is small. The original idea was to collect data all the time, even when the LED mode is not the active one. However, for the plugin to display useful information, you need to have fairly long typing sessions without rebooting the keyboard. This makes the always-collect goal much less important, because many people move with their laptops, or shut down / hibernate / etc their computers often enough to lose the collected data.

So, in a sense, the Heatmap plugin is most useful *while you're looking at it* anyway, while it is active, because the data it collects otherwise is neither persistent, nor particularly much, most likely. For those who want real heatmap data, host-side collection and analytics are far superior.

For this reason, I think it's ok if the Heatmap plugin becomes transient, and only collects data while it is active.

Reply:
> For those who want real heatmap data, host-side collection and analytics are far superior.

…unless you want information about keypresses that don't get sent to the host (e.g. layer shift keys). And don't want spurious information about virtual keys (e.g. `LSHIFT(Key_X)`).

I, for one, leave my Model01 powered for many days at a time, but I neither want the LEDs on while I'm typing, nor would I want them to stay on overnight. I forget if HostPowerManagement changes the running LED mode, but if it does, that would destroy collection for someone who wants the _keyswitch_ data, rather than the _keycode_ data.

Reply:
> …unless you want information about keypresses that don't get sent to the host (e.g. layer shift keys)

There are still options for host-side collection (KeyLogger + custom script, for example), though not ready-made.

> I forget if HostPowerManagement changes the running LED mode

It doesn't, not by default.

I too, leave the keyboard powered for many days, but my vague feeling is that we're the outliers. Nevertheless, *if* there is a need, we can add a static variant then. Until then, I think it's more practical to have it transient.